### ISSUE TITLE
dependency() static keyword is not handled according to documentation

### ISSUE THREAD

**Issue by sarum9in** [mesonbuild/meson#3084]
From documentation:
```text
`static` tells the dependency provider to try to get static
libraries instead of dynamic ones (note that this is not supported
by all dependency backends)
```

In my opinion there is either a bug or documentation is not accurate, try implies that it will prioritize static over dynamic but will still find dynamic (at least in my understanding), however:
- pkg-config dependencies fail if no static library is found.
- boost ignores it and just provides dynamic one, but this is another #3082
- python3 dependency fails if no static library is found

Should we rather say
```
`static` tells the dependency provider to use static libraries only
```
or make some them fallback to dynamic if it fails?

**Comment by nirbheek** [comment]
Yeah, like you said the situation here is quite crap because not all dependency types have the same behaviour. The intended behaviour is indeed that only static libraries will be used, but that's not what happens sometimes.

Please submit a PR to fix the documentation like you said. Not sure what we can do about Boost, we don't have a single person in charge for the dependency type, so knowledge is spread out among different people, not all of whom (can) review every Boost PR and things keep breaking.

A good first step would be add more tests (as always) since tests are automated knowledge.

A good second step would be for someone to take ownership of the boost dependency. Would you (or someone else) like to try to commit to that?

**Comment by sarum9in** [comment]
@nirbheek I am afraid that I don't have enough experience with boost on platforms different from Linux, so my knowledge of how people distribute and develop with boost is limited. I don't mind however trying to accumulate this knowledge and improve dependency type. Let me know what do you think and how you see this cooperation working.

I will submit PR for documentation.
As for the fix, #328 is a blocker for #3082. I will take a look if I can find a way to get #328 fixed...

**Comment by nirbheek** [comment]
> I am afraid that I don't have enough experience with boost on platforms different from Linux, so my knowledge of how people distribute and develop with boost is limited

I am pretty sure that no one understands all the boost platform quirks, it's the sort of thing you pick up with time and experience, and by reading the issues/PRs that people open. It's just that both I and @jpakkane are stretched thin, so we can't give Boost the time/attention it needs.

**Comment by sarum9in** [comment]
@nirbheek sounds good to me.

**Comment by nirbheek** [comment]
> I will submit PR for documentation.

I think we should wait till https://github.com/mesonbuild/meson/issues/2785 is resolved; specifically see [my comment](https://github.com/mesonbuild/meson/issues/2785#issuecomment-366798099) on it which would change the default behaviour. The feature is currently unusable with most pkg-config dependencies because of things like `-lm` which don't have static counterparts at all.

**Comment by textshell** [comment]
#2816 also proposes a way to handle selection of static vs shared linking via `default_link` and `static_paths`.
